util/conditions/patch_test.go

```diff
@@ -27,6 +27,106 @@ import (
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
+func TestNewPatchOperations(t *testing.T) {
+	fooTrue := TrueCondition("foo")
+	fooFalse := FalseCondition("foo", "reason foo", clusterv1.ConditionSeverityInfo, "message foo")
+
+	tests := []struct {
+		name    string
+		before  Getter
+		after   Getter
+		want    Patch
+		wantErr bool
+	}{
+		{
+			name:    "nil before return error",
+			before:  nil,
+			after:   getterWithConditions(),
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name:   "nil after return error",
+			before: getterWithConditions(),
+			after:  nil,
+			want:   nil,
+		},
+		{
+			name:   "nil Interface after return error",
+			before: getterWithConditions(),
+			after:  nilGetter(),
+			want:   nil,
+		},
+		{
+			name:   "nil Interface before return error",
+			before: getterWithConditions(),
+			after:  nilGetter(),
+			want:   nil,
+		},
+		{
+			name:   "No changes return empty patch",
+			before: getterWithConditions(),
+			after:  getterWithConditions(),
+			want:   nil,
+		},
+
+		{
+			name:   "No changes return empty patch",
+			before: getterWithConditions(fooTrue),
+			after:  getterWithConditions(fooTrue),
+			want:   nil,
+		},
+		{
+			name:   "Detects AddConditionPatch",
+			before: getterWithConditions(),
+			after:  getterWithConditions(fooTrue),
+			want: Patch{
+				{
+					Before: nil,
+					After:  fooTrue,
+					Op:     AddConditionPatch,
+				},
+			},
+		},
+		{
+			name:   "Detects ChangeConditionPatch",
+			before: getterWithConditions(fooTrue),
+			after:  getterWithConditions(fooFalse),
+			want: Patch{
+				{
+					Before: fooTrue,
+					After:  fooFalse,
+					Op:     ChangeConditionPatch,
+				},
+			},
+		},
+		{
+			name:   "Detects RemoveConditionPatch",
+			before: getterWithConditions(fooTrue),
+			after:  getterWithConditions(),
+			want: Patch{
+				{
+					Before: fooTrue,
+					After:  nil,
+					Op:     RemoveConditionPatch,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			got, err := NewPatchOperations(tt.before, tt.after)
+			if tt.wantErr {
```

Comment:
we should also validate the tt.wantError = false case (then we would have noticed that wantErr is missing a few times above :) )